### PR TITLE
refactor(platform)!: document creation/update/deletion does not refetch contract

### PR DESCRIPTION
**Pull request by QuantumExplorer**
## Issue being fixed or feature implemented
Document creation, update and deletion all told drive about the contract by setting the contract id. Drive would then refetch the contract from disk.

Even though most of the time it would fetch it from cache, this still was not ideal, because the processing cost of a fetch from cache is the same as a fetch from disk.

## What was done?
A new type was created that encapsulates all types on data contract info, and we made the document operations in drive more straightforward hiding away most of the complexity. Now there are just a single operation allowed for adding, updating, or deleting.

## How Has This Been Tested?
Tests continue to pass.

## Breaking Changes
Since we no longer fetch the contract in drive, drive will report less processing costs, this means that this is a breaking change.

## Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added or updated relevant unit/integration/functional/e2e tests
- [X] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [X] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

File: packages/rs-drive/src/drive/batch/drive_op_batch/document.rs [hunk 1]
```diff
@@ -51,113 +51,45 @@ pub struct DocumentOperationsForContractDocumentType<'a> {
...
         /// Document type
-        document_type_name: Cow<'a, String>,
+        document_type_info: DocumentTypeInfo<'a>,
         /// Should we override the document if one already exists?
         override_document: bool,
     },
-    /// Adds a withdrawal document.
-    AddWithdrawalDocument {
+    /// Updates a document and returns the associated fee.
+    UpdateDocument {
```
Comment (PastaPastaPasta) [inline packages/rs-drive/src/drive/batch/drive_op_batch/document.rs:66]: wym by "and returns"...? this is a stuct no? how does it "return"?

File: packages/rs-drive/src/drive/document/delete/mod.rs [hunk 2]
```diff
@@ -1,32 +1,3 @@
...
-// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
-// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
-// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
-// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
-// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
-// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
-// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
-// DEALINGS IN THE SOFTWARE.
-//
-
```
Comment (PastaPastaPasta) [inline packages/rs-drive/src/drive/document/delete/mod.rs:29]: nit: unrelated

File: packages/rs-drive/src/drive/document/delete/remove_document_from_primary_storage/v0/mod.rs [hunk 3]
```diff
@@ -24,7 +25,7 @@ impl Drive {
     #[inline(always)]
     pub(super) fn remove_document_from_primary_storage_v0(
         &self,
-        document_id: [u8; 32],
+        document_id: Identifier,
```
Comment (PastaPastaPasta) [inline packages/rs-drive/src/drive/document/delete/remove_document_from_primary_storage/v0/mod.rs:28]: nit: have these `[u8; 32]` -> `Identifier` in a separate commit to improve reviewability

File: packages/rs-drive/src/drive/object_size_info/mod.rs [hunk 4]
```diff
@@ -1,37 +1,9 @@
...
-// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
-// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
-// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
-// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
-// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
-// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
-// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
-// DEALINGS IN THE SOFTWARE.
-//
-
```
Comment (PastaPastaPasta) [inline packages/rs-drive/src/drive/object_size_info/mod.rs:29]: nit: unrelated

**Review by PastaPastaPasta — APPROVED** [review]
light utACK; generally seems to be a fine and correct refactor; there are a few things I wish were done in other PRs to make this more trivial to review. I find it is unlikely there are issues introduced here that wouldn't be caught by tests.

See comments

File: packages/rs-drive/src/drive/document/delete/remove_document_from_primary_storage/v0/mod.rs (hunk 3 above)
Comment (QuantumExplorer) [reply packages/rs-drive/src/drive/document/delete/remove_document_from_primary_storage/v0/mod.rs:28]: Indeed, but I'm lazy ><
